airbyte-cdk/python/airbyte_cdk/test/http/mocker.py

```diff
@@ -0,0 +1,86 @@
+        where the same request is added twice (in which case there will always be an exception because we will never match the second
+        request) or in a case like this:
+        ```
+        http_mocker.get(HttpRequest(_A_URL, headers={"less_granular": "1", "more_granular": "2"}), <...>)
+        http_mocker.get(HttpRequest(_A_URL, headers={"less_granular": "1"}), <...>)
+        requests.get(_A_URL, headers={"less_granular": "1", "more_granular": "2"})
+        ```
+        In the example above, the matcher would match the second mock as requests_mock iterate over the matcher in reverse order (see
+        https://github.com/jamielennox/requests-mock/blob/c06f124a33f56e9f03840518e19669ba41b93202/requests_mock/adapter.py#L246) even
+        though the request sent is a better match for the first `http_mocker.get`.
+        """
+        if isinstance(responses, HttpResponse):
+            responses = [responses]
+
+        matcher = HttpRequestMatcher(request, len(responses))
+        self._matchers.append(matcher)
+        self._mocker.get(
+            requests_mock.ANY,
+            additional_matcher=self._matches_wrapper(matcher),
+            response_list=[{"text": response.body, "status_code": response.status_code} for response in responses],
+        )
+
+    def _matches_wrapper(self, matcher: HttpRequestMatcher) -> Callable[[requests_mock.request._RequestObjectProxy], bool]:
+        def matches(requests_mock_request: requests_mock.request._RequestObjectProxy) -> bool:
+            # query_params are provided as part of `requests_mock_request.url`
+            http_request = HttpRequest(requests_mock_request.url, query_params={}, headers=requests_mock_request.headers)
+            return matcher.matches(http_request)
+
+        return matches
+
+    def __call__(self, f):  # type: ignore  # trying to type that using callables provides the error `incompatible with return type "_F" in supertype "ContextDecorator"`
+        @functools.wraps(f)
+        def wrapper(*args, **kwargs):  # type: ignore  # this is a very generic wrapper that does not need to be typed
+            with self:
+                kwargs["http_mocker"] = self
+                try:
+                    result = f(*args, **kwargs)
+                    self._validate_all_matchers_called()
+                    return result
+                except requests_mock.NoMockAddress as no_mock_exception:
+                    matchers_as_string = "\n\t".join(map(lambda matcher: str(matcher.request), self._matchers))
+                    raise ValueError(
+                        f"No matcher matches {no_mock_exception.args[0]}. Matchers currently configured are:\n\t{matchers_as_string}"
+                    ) from no_mock_exception
+                except AssertionError:
+                    try:
+                        self._validate_all_matchers_called()
```

Comment:
In the original try we attempt to call `self._validate_all_matchers_called()` and if this throws an `AssertionError`. We catch that and then call `self._validate_all_matchers_called()` again and here we try to catch a `ValueError`. 

What I'm slightly confused about is why on the first invocation we can get a `AssertionError`, but the second time we can expect a `ValueError`? And also just in general why do we need to call `self._validate_all_matchers_called()` twice? Would this not work:
```
try:
    result = f(*args, **kwargs)
    self._validate_all_matchers_called()
    return result
except requests_mock.NoMockAddress as no_mock_exception:
    matchers_as_string = "\n\t".join(map(lambda matcher: str(matcher.request), self._matchers))
    raise ValueError(
        f"No matcher matches {no_mock_exception.args[0]}. Matchers currently configured are:\n\t{matchers_as_string}"
    ) from no_mock_exception
except ValueError as http_mocker_exception:
    raise ValueError(http_mocker_exception) from None
```

I wouldn't say this is blocking ✅ , but I think this is me not understanding this block fully

Reply:
Following a sync, we did rewrite this part of the code in order to make it more readable